Automattic/jetpack
# Add filter to set custom cache timeouts for podcasts

Related to https://github.com/Automattic/wp-calypso/issues/50740

#### Changes proposed in this Pull Request:

* This PR adds a new `jetpack_podcast_feed_cache_timeout` filter to allow for custom cache timeouts for podcasts instead of the default of ([12 hours in `WP_Feed_Cache_Transient`](https://github.com/WordPress/wordpress-develop/blob/7932ebc2796c5623c01b783da1eee1253bd99da0/src/wp-includes/class-wp-feed-cache-transient.php#L36)).
* The default podcast cache timeout is undefined so we don't trigger unexpected changes in RSS feed timeouts if sites have non-default cache timeouts for RSS feeds. We only hook into the WordPress Core `wp_feed_cache_transient_lifetime` filter when we have a podcast cache timeout value returned from the filter.
  - My concern here is that if someone has hooked into the `wp_feed_cache_transient_lifetime` filter and is setting the RSS cache timeout to 14 hours, us introducing a new default could result in podcasts now having a 12 hour cache timeout (depending on the specifics about how the filters were added).
  - As noted in the filter documentation, the code only works if the site is using `WP_Feed_Cache_Transient` for caching, or their cache implementation uses the `wp_feed_cache_transient_lifetime` filter.

#### Other information:

  * No, the logic here falls around the WordPress Core `fetch_feed()` function, and I don't know how to stub that code out effectively.

#### Jetpack product discussion



This is a small enhancement to make it possible for podcast cache timeouts to be reduced so updates to podcasts can be picked up sooner.

#### Does this pull request change what data or activity we track or use?



The PR does not include tracking or data changes.

#### Testing instructions:






##### Testing on a self-hosted Jetpack site

* Ensure that you have the Jetpack plugin installed, connected, and activated
* Ensure that you have the Jetpack Beta plugin installed and activated
* Ensure that you have the Gutenberg plugin installed and activated
* Navigate to Jetpack -> Jetpack Beta in the left nav
* Enter `add/podcast-player-custom-cache-timeout` into the "Search for a Jetpack Feature Branch" search field, and click on the Activate button next to the branch when it appears
* Add the following code to your site, either manually or using a tool like [Code Snippets](https://wordpress.org/plugins/code-snippets/):
```php
add_filter( 'jetpack_podcast_feed_cache_timeout', function ( $cache_timeout, $feed_url ) {
	error_log( "jetpack_podcast_feed_cache_timeout filter; \$feed_url=$feed_url; \$cache_timeout=$cache_timeout" );
	return 3600;
} );

add_action( 'setted_transient', function ( $transient_name, $transient_value, $expiration ) {
	// We want to log sets to `feed_*` and `feed_mod_*`
	if ( 0 !== strpos( $transient_name, 'feed_' ) ) {
		return;
	}

	error_log( "Setting RSS feed transient $transient_name with expiration $expiration" );
}, 10, 3 );
```
* Ensure you have a way to view your PHP error log, either via file system access, or via a tool like [Error Log Monitor](https://wordpress.org/plugins/error-log-monitor/).
* Navigate to Posts -> Add New
* In the editor, add a new podcast player block
* In the input field, enter the URL of a podcast RSS feed that you haven't used previously. You can use a URL like `https://distributed.blog/category/podcast/feed/?somerandomsuffix873287` to ensure the feed hasn't been cached.
* View your PHP error log, and verify that you see three distinct entries like:
```
jetpack_podcast_feed_cache_timeout; $feed_url=https://distributed.blog/category/podcast/feed/?abcyxzedf; $cache_timeout=
Setting RSS feed transient feed_[hash] with expiration 3600
Setting RSS feed transient feed_mod_[hash] with expiration 3600
```
* Update the `jetpack_podcast_feed_cache_timeout` function above to return `$cache_timeout` instead of `3600`
* Go back to edit your post and enter a new feed URL (possibly with a mangled suffix)
* Verify that you see the following entries in the PHP error log:
```
jetpack_podcast_feed_cache_timeout; $feed_url=https://distributed.blog/category/podcast/feed/?abcyxzedf; $cache_timeout=
Setting RSS feed transient feed_[hash] with expiration 43200
Setting RSS feed transient feed_mod_[hash] with expiration 43200
```

##### Testing steps on a WordPress.com Simple site

I will document these steps as part of D83749-code, as this depends on a number of internal tools and a different RSS caching infrastructure.
